SCons/Subst.py

```diff
@@ -420,7 +420,7 @@ def func(l, conv=self.conv, substitute=self.substitute, lvars=lvars):
             # which does not work the signature module, and the Null class returns an empty
             # string if called on, so we make an exception in this condition for Null class
             if (isinstance(s, SCons.Util.Null) or
-                set(signature(s).parameters.keys()) == set(['target', 'source', 'env', 'for_signature'])):
+                    getfullargspec(s).args[:4] == ['target', 'source', 'env', 'for_signature']):
```

Comment:
Why only `:4`?  shouldn't there only be 4 arguments which match in order the list above?

Reply:
I'm only looking at the first four functions because I'm allowing for extra arguments in the function definition that may be filled by `functools.partial` or similar later on. I also don't think I expressed the problem correctly, either. The problem I ran across is that functions which would work perfectly well as generators (because of `functools.partial`) are being classified by the conditional as non-generator functions, which results in them being stringified rather than executed. What I am trying to do is allow a bit more flexibility while at the same time make problems more clear to the user.

This is just my first brush at dealing with the problem. I just noticed that the first branch of the conditional I am changing isn't exactly the opposite case of the second branch. In the first branch, we are looking *only* for free functions (and subclasses of `Scons.Util.Null`, and in the second branch we are assuming the callable is an object for which `__str__` is always meaningful, which is not actually the case. I'm exploring some ideas for how we might avoid that assumption as well.

How do you feel about introducing a template class that users can inherit from which will raise `Scons.Error.UserError` when `__str__` is called on it directly (rather than the return value of the generator method)? I can show you what I mean in my next commit if you are interested in it.

Reply:
So increase the amount of python knowledge needed to create a generator function?

Reply:
curious why getfullargspec? from the python docs for https://docs.python.org/3/library/inspect.html#inspect.getfullargspec:

> Note that signature() and Signature Object provide the recommended API for callable introspection, and support additional behaviours (like positional-only arguments) that are sometimes encountered in extension module APIs. This function is retained primarily for use in code that needs to maintain compatibility with the Python 2 inspect module API.

Reply:
The defined API is that the function (or the callable) should take only those four arguments.

Reply:
I'm suggesting we allow users with more advanced Python knowledge to use it with more advanced Python. The base level of knowledge required to create a working generator function remains exactly the same, and I am hoping to make that even better by emitting a useful error rather than doing unexpected behavior. With the changes I am proposing so far, the callable that ends up being used in a `Subst` *does* only take those four arguments because all the others will have been filled.

The problem I am identifying is that inspecting the callable's signature the way we are "sees through" the partial function to the original signature, which is what is causing the mismatch and is causing the function to be stringified rather than an error being raised. If I do something wrong that the docs say won't work, what happens is that I get unexpected behavior rather than an error that tells me what I did wrong. In fact, the partial function would have worked just fine, and it was only the overly strict signature check that caused unexpected behavior.

Being as strict as we are with this signature check is disenfranchising the advanced Pythonistas who already know what they're doing and isn't helping the less advanced users to know what they got wrong. We can do better checking without this level of strictness. Besides, `inspect.getargspec` is deprecated as of Python 3.0 anyway, and `inspect.getfullargspec` is not. It also just uses the return value of `inspect.signature()` as compatibility with the Python 2 API. If you prefer to cut the cord entirely with Python 2 in all new code, I can just use `inspect.signature()` directly.

Reply:
SCons is (and has been) Py 3.5+ only since 4.0...

Reply:
How would such get called by Subst() (with any more the the API specified args)?  Since that's where this code is being used..

Reply:
As an example related to something I was working on, let's say I have a function I would like to use as a generator, but I want to conditionally output some debugging messages:

```
def some_generator(source, target, env, for_signature, debug=False):
    if debug:
        print("Some debugging message")
   ...
```

I should be able to use that as a generator anywhere with no problems because the `debug` keyword argument is defaulted. If I wanted to give that as a generator in a `Subst`, I might be tempted to try something like this (where `debug` is a boolean):

```
env['SOME_GENERATOR'] = functools.partial(some_generator, debug=debug)
```

However, if I have a `Subst` that references "$SOME_GENERATOR", I get a confusing result because the original function has extra arguments. This is true even if I *don't* use `functools.partial` for this function. As an experienced Python developer, I would expect the function to work even without using `functools.partial`, but it doesn't do that, either. Instead, I always get a `repr` of `some_generator` in my `Subst` result regardless of the options I pass.

Any reasonably experienced Python developer coming into SCons anew is going to be utterly perplexed as to why this is happening because we're used to functional-style programming. And in fact, I can make this work by simply *not* using `functools.partial` and either generating a function with the extra arguments in a closure or just using a `lambda`. It's only `functools.partial` - the cleanest and most "functional" way - that doesn't work. A novice programmer who simply gets it wrong, maybe with a typo in one of the argument names, is just going to be confused by why it's doing strange things. I would like to make this situation better for both the novices and the experienced Python developers if I can.

Reply:
I think there are two separate things to think about here.

- First, the old code didn't enforce order for the four "key" arguments. The new code does. I think that is probably a worthwhile improvement. I have no opinion on the exact mechanism for that (`argspec` vs `getfullargspec`, etc). I think we should do whatever is best given the python version support constraints that are in play.

- Second, the old code would not invoke the generator if the function was something that took, say, 5 arguments, with the first four having the correct names and in the correct order, and the 5th defaulted. That's still a function that Subst could meaningfully call with the four arguments.

In other words:

```
>>> def foo(target, source, env, for_signature, debug=False):
...     print(target, source, env, for_signature, debug)
...
>>>
>>> foo('t', 's', 'e', 'fs')
t s e fs False
>>> foo('t', 's', 'e', 'fs', True)
t s e fs True
```

If you had `env['FOO'] = foo`, that should work just fine as a generator: SCons doesn't need to care about anything that comes after the fourth argument. As long as the first four function arguments have the correct names and are in the correct order, SCons should call it as a generator.

If calling it that way fails (perhaps the 5th argument isn't actually defaulted!), that's on the user who wrote the generator. They will get an error from the python runtime telling them that the function was called with four arguments but requires 5.

Reply:
> First, the old code didn't enforce order for the four "key" arguments. The new code does. I think that is probably a worthwhile improvement.

It didn't enforce order because scons is setting the keyword args directly when calling the callable, so does order matter?

Reply:
> Second, the old code would not invoke the generator if the function was something that took, say, 5 arguments, with the first four having the correct names and in the correct order, and the 5th defaulted. That's still a function that Subst could meaningfully call with the four arguments.

Okay so for this situation, how would you ever have the function run with something other than the default value? If thats the case it could just be hard coded in the function call instead.

Reply:
Indeed.. It's a bit contrived.

Unless you were going to call the function elsewhere than via subst()
Still I think we're somewhat deep in the weeds here.

Reply:
> Unless you were going to call the function elsewhere than via subst()

Yeah its a possible case, but seems like your coupling different code and in general not a good idea. Like what use case is there where you want something to run in subst context and then also run in some other non subst context? Anyways I am not against allowing more args, just trying to find reason.

Reply:
This has nothing to do with running a function in multiple ways than just a `subst()`, and if I have left anyone with the impression that's what I was describing it was unintentional. I said I have a function that is used in `subst()` calls, and the extra arguments are about preventing duplication of code between similar calls. In other words, rather than make multiple similar functions that differ only in the information they need to run - as in, which data to process or whether an optional behavior is performed. The mention of `functools.partial` in the original post was just a demonstration of the point from the code I am working with.

There are ways I could get around this right now, such as generating a closure or storing state in the construction environment, but the simplest and most obvious way to an experienced Python developer is to just allow the function to accept extra arguments in the first place and make sure they are filled before the call happens. The strict requirement actually hampers our ability to determine user intent, so we end up doing the wrong thing from the user's perspective in ways that are hard for them to diagnose on their own.

As @acmorrow has pointed out, we can just let Python tell users they have unfilled arguments and be satisfied that the arguments we need to run a `subst()` are present, and - if the call succeeds - all arguments been filled. The reason for adding order checking is so we can be sure the user intended a generator function because it matches the required signature in the documentation. SCons doesn't really need to care about whether *only* the required arguments have been given if the call succeeds, and if it doesn't, Python can handle telling the user what happened.

Reply:
> I said I have a function that is used in subst() calls, and the extra arguments are about preventing duplication of code between similar calls.

There's already API for this. Here is the example from the documentation modified for the debug flag use case you mentioned:

```
import SCons

env = Environment()

class my_generator:
    def __init__(self, debug):
        self.debug = debug

    def __call__(self, env, source, target, for_signature):
        if self.debug and not for_signature:
            print(f"DEBUG: {target}")
        return "echo $TARGET"

debug = True
env["MY_GEN"] = my_generator
env["MY_SUBST1"] = "${MY_GEN(%s)}" % debug
env["MY_SUBST2"] = "${MY_GEN(%s)}" % (not debug)

env.Command('not_built1', [], SCons.Action.Action("$MY_SUBST1", None))
env.Command('not_built2', [], SCons.Action.Action("$MY_SUBST2", None))
```

outputs:
```
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
DEBUG: not_built1
not_built1
not_built2
scons: done building targets.
```

Reply:
A side note, I would like to see the functools.partial be a useable option with subst function. It makes it simpler to write the above example. Not sure what constaints there should be or if this PR is the one to do it but it would be an nice feature.